File: python/largest_rectangle_histogram.py
class Solution(object):
    # O(n) time | O(n) space
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        maxArea = 0
        stack = []

        for i, h in enumerate(heights):
            idx = i
            while len(stack) != 0 and stack[-1][1] > h:
                curr = stack.pop()
                currArea = (i - curr[0]) * curr[1]
                idx = curr[0]
                maxArea = max(maxArea, currArea)
            stack.append((idx, h))
        
        while len(stack) != 0:
            curr = stack.pop()
            currArea = (len(heights) - curr[0]) * curr[1]
            maxArea = max(maxArea, currArea)
        
        return maxArea

solution = Solution()
print(solution.largestRectangleArea([2,1,5,6,2,3]))
            

File: src/main/java/august_2025/LargestRectangleHistogram.java
package august_2025;


import java.util.Stack;

public class LargestRectangleHistogram {

    public static void main(String[] args) {
        int[] heights = {2,1,5,6,2,3};

        LargestRectangleHistogram largestRectangleHistogram = new LargestRectangleHistogram();
        int result = largestRectangleHistogram.largestRectangleArea(heights);
        System.out.println(result);

    }

    // O(n) time | O(n) space
    public int largestRectangleArea(int[] heights) {
        int maxArea = 0;
        Stack<int[]> stack = new Stack<>();
        for (int i = 0; i < heights.length; i++) {
            int curr = heights[i];
            int idx = i;
            while (!stack.isEmpty() && stack.peek()[0] >= curr) {
                int[] pop = stack.pop();
                int area = pop[0] * (i - pop[1]);
                maxArea = Math.max(maxArea, area);
                idx = pop[1];
            }
            stack.push(new int[] {curr, idx});
        }

        while (!stack.isEmpty()) {
            int[] pop = stack.pop();
            int area = pop[0] * (heights.length - pop[1]);
            maxArea = Math.max(maxArea, area);
        }

        return maxArea;
    }

}
